src/games/strategy/triplea/delegate/MoveValidator.java

```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
    */
   public static String validateCanal(final Territory territory, final Route route, final Collection<Unit> units,
       final PlayerID player, final GameData data) {
+    String failureMessage = null;
     final Set<CanalAttachment> canalAttachments = CanalAttachment.get(territory);
-    if (canalAttachments.isEmpty()) {
-      return null;
-    }
-    final Iterator<CanalAttachment> iter = canalAttachments.iterator();
-    while (iter.hasNext()) {
-      final CanalAttachment attachment = iter.next();
-      if (attachment == null) {
+    for (final CanalAttachment canalAttachment : canalAttachments) {
+      if (!needToCheckCanal(canalAttachment, route, data)) {
```

Comment:
Line 1428 doesn't make sense. If we do not need to check canal, then we check if we can pass through the canal. Is 1428 supposed to have the "!" in the if statement?